crates/ide/src/runnables.rs

```diff
@@ -575,6 +616,120 @@ fn has_test_function_or_multiple_test_submodules(
+        if self.expect_test {
+            builder.push("Expect");
+        }
+        if self.insta {
+            builder.push("Insta");
+        }
+        if self.snapbox {
+            builder.push("Snapbox");
+        }
+
+        let res: SmolStr = builder.join(" + ").into();
+        if res.is_empty() {
+            None
+        } else {
+            Some(format_smolstr!("↺\u{fe0e} Update Tests ({res})"))
+        }
+    }
+
+    pub fn env(&self) -> SmallVec<[(&str, &str); 3]> {
+        let mut env = SmallVec::new();
+        if self.expect_test {
+            env.push(("UPDATE_EXPECT", "1"));
+        }
+        if self.insta {
+            env.push(("INSTA_UPDATE", "always"));
+        }
+        if self.snapbox {
+            env.push(("SNAPSHOTS", "overwrite"));
+        }
+        env
+    }
+}
+
+impl<'a, 'b> TestDefs<'a, 'b> {
+    fn new(
+        sema: &'a Semantics<'b, RootDatabase>,
+        current_krate: hir::Crate,
+        file_range: hir::FileRange,
+    ) -> Self {
+        Self(sema, current_krate, file_range)
+    }
+
+    fn update_test(&self) -> UpdateTest {
+        UpdateTest { expect_test: self.expect_test(), insta: self.insta(), snapbox: self.snapbox() }
+    }
+
+    fn expect_test(&self) -> bool {
+        self.find_macro("expect_test", &["expect", "expect_file"])
+    }
+
+    fn insta(&self) -> bool {
+        self.find_macro(
+            "insta",
+            &[
+                "assert_snapshot",
+                "assert_debug_snapshot",
+                "assert_display_snapshot",
+                "assert_json_snapshot",
+                "assert_yaml_snapshot",
+                "assert_ron_snapshot",
+                "assert_toml_snapshot",
+                "assert_csv_snapshot",
+                "assert_compact_json_snapshot",
+                "assert_compact_debug_snapshot",
+                "assert_binary_snapshot",
+            ],
+        )
+    }
+
+    fn snapbox(&self) -> bool {
+        self.find_macro("snapbox", &["assert_data_eq", "file", "str"])
+    }
+
+    fn find_macro(&self, crate_name: &str, paths: &[&str]) -> bool {
+        let db = self.0.db;
+
+        let Some(dep) =
+            self.1.dependencies(db).into_iter().find(|dep| dep.name.eq_ident(crate_name))
+        else {
+            return false;
+        };
+        let module = dep.krate.root_module();
+        let scope = module.scope(db, None);
+
+        paths
+            .iter()
+            .filter_map(|path| {
+                let (_, def) = scope.iter().find(|(name, _)| name.eq_ident(path))?;
+                match def {
+                    hir::ScopeDef::ModuleDef(hir::ModuleDef::Macro(it)) => Some(it),
+                    _ => None,
+                }
+            })
+            .any(|makro| {
+                Definition::Macro(*makro)
+                    .usages(self.0)
+                    .in_scope(&SearchScope::file_range(self.2))
+                    .at_least_one()
+            })
+    }
```

Comment:
instead of `scope.iter()` checking the parts we should just use `Semantics::resolve_mod_path` and construct `ModPath`s from the string paths, that way we re-use our normal resolution logic. And again this can be cached like the other part at the start of the runnables stuff.

Reply:
Currently we are unable to create `ModPath` directly from string paths because we are unable to construct `Name` from a String. An alternative option is to use `new_symbol_root`, but I am unsure if it is the correct approach.

Reply:
That's fine, you can use `new_symbol_root` here. We don't care about hygiene for this.